Fixes the getErrorWithSeverity function.

luciansmith opened the pull request:
When the error in question is an XMLError and not an SBMLError, the old dynamic_cast fails and returns NULL.  This is I think because the run-time tests that dynamic_cast try to run cannot run because XMLError has no virtual functions, and the test cannot be run, and dynamic_cast returns NULL.  Changing to a static_cast works instead, because it has no checks.

luciansmith committed:
Fixes the getErrorWithSeverity function.

When the error in question is an XMLError and not an SBMLError, the old dynamic_cast fails and returns NULL.  This is I think because the run-time tests that dynamic_cast try to run cannot run because XMLError has no virtual functions, and the test cannot be run, and dynamic_cast returns NULL.  Changing to a static_cast works instead, because it has no checks.

skeating reviewed (changes requested):
This looks like its breaking an existing test in the sbml dir
sorry dont have time to look 
hopefully be freer by end of week

fbergmann commented on src/sbml/SBMLErrorLog.cpp:
```diff
@@ -385,7 +385,7 @@ SBMLErrorLog::getErrorWithSeverity(unsigned int n, unsigned int severity) const
     if (matcher(*it))
     {
-      if (count == n) return dynamic_cast<const SBMLError*>(*it);
+      if (count == n) return static_cast<const SBMLError*>(*it);
```
this cannot work, since the iterator could be pointing to an error of type `XMLError`, which is the base class of SBML error. So the dynamic cast was there to return a NULL error in that case. The static cast would just force the error object into an SBML error, and create a bad object. 

Thus the test failures: 

```

2021-10-01T19:34:00.9150270Z 35:  Generic Attribute Functions
2021-10-01T19:34:00.9655721Z 35: 99%: Checks: 1737, Failures: 1, Errors: 0
2021-10-01T19:34:00.9657744Z 35: D:\a\libsbml\libsbml\src\sbml\test\TestSBMLError.cpp:243:F:SBMLError:test_SBMLErrorLog_getErrorWithSeverity_XMLError:0: Assertion 'error->getErrorId() == XMLContentEmpty' failed
2021-10-01T19:34:00.9695739Z  35/184 Test  #35: test_sbml_sbml_run ............................***Failed    2.04 sec
2021-10-01T19:34:00.9700201Z test 36
```

fbergmann commented on src/sbml/test/TestSBMLError.cpp:
```diff
@@ -173,6 +175,76 @@ START_TEST(test_SBMLErrorLog_removeAll)
+    SBMLDocument* doc = readSBMLFromString("Not actually SBML.");
+
+    const SBMLError* error = doc->getErrorWithSeverity(0, LIBSBML_SEV_ERROR);
+    fail_unless(error != NULL);
```
I dont think this should pass.

luciansmith commented:
In that case, 'getError' should fail, then, since it uses static_cast.  But nobody's complained about that, and the static_cast works for me on Windows, so I'm not sure what the correct solution is.

luciansmith committed:
Upcast every added XMLError.

Since we return SBMLErrors and not XMLErrors from the error log, we must ensure that all errors are indeed SBMLErrors.

This allows the dynamic casting of all errors from the getError and getErrorWithSeverity functions.

Note that if any XML error had something in the message beyond the basic 'this is what's added for this error code', it will be lost, since there's no way to split that message from the basic message.  If the SBMLError constructor was given the 'getMessage' from the XMLError, the basic message would be duplicated.

Also fix a couple warnings while I'm at it.

luciansmith committed:
Fix the message issue; also test 'getError'.

luciansmith commented:
OK, tried the upconverting thing.  Test work (again) on Windows; hopefully they'll work elsewhere too.

luciansmith committed:
Error code check fix.

Other parsers apparently return different error codes, so just check to see if it's *an* XML error code; and don't worry about which one.

luciansmith committed:
Update CMakeLists.txt

Argh; don't know how those changes were committed...

luciansmith committed:
Update CMakeLists.txt

spacing to match original exactly.

fbergmann commented on src/sbml/xml/XMLErrorLog.cpp:
```diff
@@ -126,6 +126,14 @@ XMLErrorLog::add (const XMLError& error)
+  const SBMLError* serror = dynamic_cast<const SBMLError*>(&error);
+  if (serror == NULL)
+  {
+      SBMLError sbmlerr(error.getErrorId(), 0, 0, "", error.getLine(), error.getColumn(), error.getSeverity(), error.getCategory());
```
@luciansmith I didnt' think you would add the upconversion here. This is potentially problematic, as sarah was working on liblx, which essentially removes the XML layer from libSBML (and all references to SBML), so i was expecting you to add this functionality at the 'getError' level of the sbml document, for the cases where you get a non-sbml error returned. but @skeating  will have to clearify.

luciansmith replied:
If 'getError' returned an object, it would be easy to upconvert there, but since it returns a pointer, you need to save it somewhere.  I suppose it could delete and re-add the object at that point?

Another option might be to make 'add' virtual, and override it with SBMLError?  The problem I was running into was that I couldn't access member variables that way.

luciansmith replied:
Done!  Moved the upcast to SBMLErrorLog's 'add', by virtue of making XMLErrorLog's 'add' virtual, and adding a 'setMessage' function to SBMLError (protected) and setting the SBMLErrorLog as a friend.

luciansmith committed:
Move the upcast to SBMLErrorLog

Had to make the 'add' functions virtual on XMLErrorLog to do this, and add a 'setMessage' function to SBMLError, since otherwise the member variable wasn't available.  But this moves all the SBML-specific functionality to SBML, and leaves the XML alone (other than making things virtual).

luciansmith committed:
Merge branch 'fix-getErrorWithSeverity' of https://github.com/sbmlteam/libsbml into fix-getErrorWithSeverity

luciansmith committed:
No longer need this include.

fbergmann commented on src/sbml/SBMLErrorLog.cpp:
```diff
@@ -172,6 +172,27 @@ SBMLErrorLog::add (const SBMLError& error)
+    {
+        SBMLError sbmlerr(error.getErrorId(), 0, 0, "", error.getLine(), error.getColumn(), error.getSeverity(), error.getCategory());
+        sbmlerr.setMessage(error.getMessage());
+        return XMLErrorLog::add(sbmlerr);
```
that return statement looks odd ... when looking at the void method. Might be better to have it after the add statement,

luciansmith replied:
I dunno; I feel like it's a pretty common idiom.  We're calling the base class version of the function and doing the same thing.  This also doesn't have to be changed in the future if we decide to (say) change the signature so it returns an error code on failure.

luciansmith replied:
(It's obviously fine to expand it to two lines if anyone really wants it, if course.)

fbergmann commented on src/sbml/SBMLError.h:
```diff
@@ -1253,6 +1253,10 @@ class LIBSBML_EXTERN SBMLError : public XMLError
 
+  virtual void setMessage(const std::string& message);
+
+  friend class SBMLErrorLog;
```
why would we need the friend statement, just because setMessage is private? I think i actually wouldn't mind a public `setMessage`, as then people could customize error messages more easily.

luciansmith replied:
This follows the current design of XMLError, which is that it cannot be changed after construction.  This has always seemed like an odd decision to me, but it's baked into the design from the get go.  Here, we *have* to allow message editing, because you cannot construct the object with the entire message; the automatic message will always be added first, so this allows for that situation while making the smallest design change to XMLError.

I would be all for changing the design so that XMLError itself just made public 'setX' functions for everything, but that's a different issue.

skeating replied:
Actually that decision even predates me arrgg

Yes I'm happy to with allowing Errors to be edited after creation - this allows people flexibility.
So pubic setMessage/setSeverity

:-)

luciansmith replied:
In that case, I'm inclined to go ahead and change XMLError itself--will that mess with anything on the libx side of things?

fbergmann replied:
@luciansmith i think that is an improvement, that liblx will also have to mirror. Thanks for doing this.

luciansmith committed:
Update XMLError with setter functions.

This allows us to remove the one-off setter function from SBMLError, as well.

luciansmith committed:
Allow the setting of the category/severity strings.

The strings are normally automatically set when the code is set, but if it's an extended category/severity, the core setter won't know what it was.

Also made the functions virtual so they can be overwritten if need be.

fbergmann reviewed (approved):
looks all good to me

luciansmith commented:
@skeating : Does this work for you now?  You have a requested change.
